src/dfx/src/commands/canister/update_settings.rs

```diff
@@ -222,6 +232,31 @@ pub async fn exec(
     Ok(())
 }
 
+fn user_is_removing_themselves_as_controller(
+    env: &dyn Environment,
+    call_sender: &CallSender,
+    opts: &UpdateSettingsOpts,
+) -> DfxResult<bool> {
+    let caller_principal = match call_sender {
+        CallSender::SelectedId => env
+            .get_selected_identity_principal()
+            .context("Selected identity is not instantiated")?
+            .to_string(),
+        CallSender::Wallet(principal) => principal.to_string(),
+    };
+    let removes_themselves = opts
+        .remove_controller
+        .as_ref()
+        .map(|remove| remove.contains(&caller_principal))
+        .unwrap_or_default();
+    let not_in_new_controllers = opts
+        .set_controller
+        .as_ref()
+        .map(|set| !set.contains(&caller_principal))
+        .unwrap_or_default();
+    Ok(removes_themselves || not_in_new_controllers)
```

Comment:
I had a little trouble following what was going on here, I think due to the combination of `Option<Vec<_>>`, (sometimes )inverted membership in `Vec<_>`, and `unwrap_or_default`.  But both checks only have a single case in which they are `true`.

What would you think about this?
```suggestion
    let removes_themselves =
        matches!(&opts.remove_controller, Some(remove) if remove.contains(&caller_principal));
    let sets_without_themselves =
        matches!(&opts.set_controller, Some(set) if !set.contains(&caller_principal));
    Ok(removes_themselves || sets_without_themselves)
```

Reply:
I wasn't aware of `matches!` with `if`. Nice idea!